WIP: Create new document for rate-limiting issue (#3654)
thanos-io/thanos: docs/proposals/202006_thanos_grpc_rate_limiting.md

---
title: Thanos Rate Limiting for GRPC
type: proposal
menu: proposals
status: WIP
owner: bwplotka
---

### Related Tickets

* https://github.com/thanos-io/thanos/issues/3525 (main)

## Summary and motivation

> We want to implement a way to  "rate limiting" concept in thanos, which should prevent to overlaoad CPU and other resource server side which can result with high concurrent request in GRPC.

Reviewer: ```suggestion
We want to implement a way to "rate limiting" concept in Thanos, which should prevent overloading CPU and other resources on the server-side due to high concurrent requests in GRPC.
```

> At the point of this writing GRPC has no standard library way to implement a rate-limiter. Allthough there is a interceptor from Grpc ecosystem ( see later)

Reviewer: ```suggestion
At the point of this writing, gRPC has no standard library way to implement a rate-limiter. Although there is an interceptor from gRPC ecosystem ( see later)
```
Let's be consistent it's called `gRPC`

> The goal of the  mechanism should be a unified, safe way for rate limiting, both per component but also potentially per tenant.

Reviewer: I would say another goal is to reuse similar rate limit technique for HTTP as for gRPC

Author: yop right

There are basically 2 ways of solving the problem:

> We should avoid such saturation with a rate-limiting concept or investigate if client queuing is better.

Reviewer: client queuing is also rate limiting concept, no? (:

Author: lol right
